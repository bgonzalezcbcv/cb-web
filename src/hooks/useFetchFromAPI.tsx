import React, { useCallback, useEffect } from "react";
import { DefaultApiResponse } from "../core/interfaces";

export enum FetchStatus {
	Initial,
	Fetching,
	Error,
}

/**
 *
 * @param fetchFunction: Function from the APIStore needed to fetch. Tip: If you need to pass a parameter, like an ID,
 * pass this parameter as () => myAPIFunc(id)
 * @param onFetch: this is the setter that will be used when the fetch was correct to save the data.
 */
function useFetchFromAPI<Type>(
	fetchFunction: (...params: unknown[]) => Promise<DefaultApiResponse<Type>>,
	onFetch: (data: Type) => void
): { fetchStatus: FetchStatus; refetch: () => void } {
	const [fetchStatus, setFetchStatus] = React.useState(FetchStatus.Fetching);

	const fetchHandler = useCallback(async () => {
		setFetchStatus(FetchStatus.Fetching);

		const response = await fetchFunction();

		if (response.success && response.data) {
			onFetch(response.data);
			setFetchStatus(FetchStatus.Initial);
		} else setFetchStatus(FetchStatus.Error);
	}, []);

	useEffect(() => {
		fetchHandler();
	}, []);

	return {
		fetchStatus,
		refetch: () => fetchHandler(),
	};
}

export default useFetchFromAPI;
