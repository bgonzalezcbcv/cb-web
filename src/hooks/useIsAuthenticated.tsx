import { useNavigate } from "react-router-dom";

import { UserRole } from "../core/interfaces";
import { DataStore } from "../core/DataStore";
import { User } from "../core/Models";

/**
 *
 * @param authRoles: Roles that are authorized to view this component.
 * @param additionalValidation: Callback to do a custom validation, false to redirect.
 */
export default function useIsAuthenticated(authRoles: UserRole[], additionalValidation?: (loggedUser: User) => boolean): void {
	const navigate = useNavigate();
	const loggedUser = DataStore.getInstance().loggedUser;

	if (!loggedUser) return navigate("/");

	if (!authRoles.includes(loggedUser.role)) return navigate("/");

	if (additionalValidation && !additionalValidation(loggedUser)) return navigate("/");
}
