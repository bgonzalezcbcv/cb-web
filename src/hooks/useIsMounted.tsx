import { useEffect, useState } from "react";

/**
 * Use this hook to "hide" components when they are still to be mounted on the DOM. Specially useful for facilitating automated tests.
 */
export function useIsMounted(): [boolean] {
	const [mounted, setMounted] = useState(false);

	useEffect(() => setMounted(true));

	return [mounted];
}
