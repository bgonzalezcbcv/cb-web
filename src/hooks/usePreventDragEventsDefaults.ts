import React from "react";

const { useEffect } = React;

function usePreventDragEventsDefaults(): void {
	useEffect(() => {
		window.addEventListener(
			"dragover",
			function (e) {
				e = e || event;
				e.preventDefault();
			},
			false
		);

		window.addEventListener(
			"drop",
			function (e) {
				e = e || event;
				e.preventDefault();
			},
			false
		);
	});
}

export default usePreventDragEventsDefaults;
