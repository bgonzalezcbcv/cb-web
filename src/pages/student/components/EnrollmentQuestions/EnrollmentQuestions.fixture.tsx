export const initialStudentData = {
	question_categories: [
		{
			category: "categoria 1",
			questions: [
				{
					id: "1",
					question:
						"question 1 pero es mucho muchomuchomuchomucho muchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomuchomuchomucho mas larga para romper todo",
					answer: "answer 1a",
				},
				{
					id: "2",
					question: "question 2",
					answer: "answer 2",
				},
				{
					id: "3",
					question: "question 3",
					answer: "answer 3",
				},
			],
		},
		{
			category: "categoria 2",
			questions: [
				{
					id: "12",
					question: "question 12",
					answer: "answer 12",
				},
				{
					id: "22",
					question: "question 22",
					answer: "answer 22",
				},
				{
					id: "32",
					question: "question 32",
					answer: "answer 32",
				},
			],
		},
	],
};

export const expectedNewStudentData = {
	question_categories: [
		{
			category: "categoria 1",
			questions: [
				{
					id: "1",
					question:
						"question 1 pero es mucho muchomuchomuchomucho muchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomuchomuchomucho mas larga para romper todo",
					answer: "answer 1anew",
				},
				{
					id: "2",
					question: "question 2",
					answer: "answer 2",
				},
				{
					id: "3",
					question: "question 3",
					answer: "answer 3",
				},
			],
		},
		{
			category: "categoria 2",
			questions: [
				{
					id: "12",
					question: "question 12",
					answer: "answer 12",
				},
				{
					id: "22",
					question: "question 22",
					answer: "answer 22",
				},
				{
					id: "32",
					question: "question 32",
					answer: "answer 32",
				},
			],
		},
	],
};
