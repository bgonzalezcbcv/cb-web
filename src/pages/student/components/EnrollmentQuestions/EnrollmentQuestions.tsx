import _ from "lodash";
import React, { useCallback, useEffect, useState } from "react";

import {
	Accordion,
	AccordionDetails,
	AccordionSummary,
	Box,
	Button,
	FormControl,
	InputLabel,
	List,
	MenuItem,
	Select,
	SelectChangeEvent,
	Typography,
} from "@mui/material";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

import * as API from "../../../../core/ApiStore";
import { Question as QuestionModel, Student, Cicle } from "../../../../core/Models";
import Question from "./components/Question";

export interface EnrollmentQuestionsProps {
	student: Student;
	editable: boolean;
	onChange: (newStudent: Student) => void;
	viewMode: string;
}

export default function EnrollmentQuestions(props: EnrollmentQuestionsProps): React.ReactElement {
	const { student, editable, onChange, viewMode } = props;

	const isViewMode = viewMode == "VIEW";
	const cicleQuestionCategories = student.cicle_question_categories;

	const [selectedCicle, setSelectedCicle] = useState<Cicle>(Cicle.None);
	const [question_categories, setQuestionCategories] = useState(cicleQuestionCategories[0].question_categories);
	const [error, setError] = useState<string | null>(null);

	const studentCicles = Object.keys(Cicle) as Array<keyof typeof Cicle>;
	const indexOfCicle = Object.values(Cicle).indexOf(student.cicle);
	const validCicles = studentCicles.splice(0, indexOfCicle + 1);

	const onChangeHandler = (changedQuestionCategoryIndex: number, changedQuestionIndex: number, newAnswerValue: string): void => {
		const newStudentData: Student = _.cloneDeep(student);
		newStudentData.cicle_question_categories.filter((q) => q.cicle == selectedCicle)[0].question_categories[changedQuestionCategoryIndex].questions[
			changedQuestionIndex
		].answer = newAnswerValue;

		onChange(newStudentData);
	};

	useEffect(() => {
		if (error) window.setTimeout(() => setError(null), 3000);
	}, [error]);

	const handleCicleChange = (event: SelectChangeEvent): void => {
		const cicleString = event.target.value;
		const cicleList = Object.keys(Cicle) as Array<keyof typeof Cicle>;
		const cicleIndex = Object.values(Cicle).indexOf(cicleString as Cicle);
		const cicleKey = cicleList[cicleIndex];
		const cicle = Cicle[cicleKey];
		const cicleCategories = student.cicle_question_categories.filter((q) => q.cicle == cicle)[0].question_categories;

		setSelectedCicle(cicle);
		setQuestionCategories(cicleCategories);
	};

	const sendAnswers = useCallback(async (): Promise<void> => {
		const answers: QuestionModel[] = [];
		for (const category of question_categories) {
			answers.concat(category.questions);
		}
		const sendAnswersResponse = await API.postAnswersEnrollmentQuestions(student.id, indexOfCicle, answers);

		if (!sendAnswersResponse) {
			setError("Error al enviar las respuestas.");
			return;
		}
	}, [question_categories]);

	return (
		<Box>
			<Box display="flex">
				<Box
					sx={{
						display: "flex",
						justifyContent: "flex-start",
						width: "50%",
					}}>
					<FormControl variant="standard" sx={{ width: "50%" }}>
						<InputLabel>Seleccionar Ciclo</InputLabel>
						<Select value={selectedCicle} onChange={handleCicleChange}>
							{validCicles.map((key: keyof typeof Cicle) => {
								return (
									<MenuItem key={key} value={Cicle[key]}>
										{Cicle[key]}
									</MenuItem>
								);
							})}
						</Select>
					</FormControl>
				</Box>
				<Box
					sx={{
						display: "flex",
						justifyContent: "flex-end",
						width: "50%",
					}}>
					<Button sx={{ display: isViewMode ? "" : "none" }} variant="outlined" onClick={(): unknown => sendAnswers}>
						Enviar Respuestas
					</Button>
				</Box>
			</Box>
			<List>
				{question_categories.map((category, categoryIndex): React.ReactElement => {
					return (
						<div key={"category" + categoryIndex}>
							<Accordion
								sx={{
									flexDirection: "column",
									justifyContent: "space-between",
									display: "flex",
									flex: 1,
									height: "100%",
									width: "100%",
									alignContent: "center",
								}}>
								<AccordionSummary expandIcon={<ExpandMoreIcon />}>
									<Typography> {category.category} </Typography>
								</AccordionSummary>

								<AccordionDetails>
									{category.questions.map(
										(question, questionIndex): React.ReactElement => (
											<Question
												key={`${category}-${categoryIndex}-question-${questionIndex}`}
												question={question}
												questionIndex={questionIndex}
												editable={editable && student.cicle == selectedCicle}
												onChangeQuestion={(newAnswer): void => onChangeHandler(categoryIndex, questionIndex, newAnswer)}
											/>
										)
									)}
								</AccordionDetails>
							</Accordion>
						</div>
					);
				})}
			</List>
		</Box>
	);
}
