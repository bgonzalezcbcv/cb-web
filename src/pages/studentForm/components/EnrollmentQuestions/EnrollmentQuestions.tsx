import React, { useEffect, useState } from "react";
import _ from "lodash";

import { Divider, Grid, List, ListItem, TextField, Typography } from "@mui/material";

import useDebounce from "../../../../hooks/useDebounce";

export type Question = {
	id: string;
	question: string;
	answer: string;
};

export interface QuestionCategories {
	category: string;
	questions: Question[];
}

export interface EnrollmentQuestionsProps {
	studentQuestionCategories: QuestionCategories[];
	editable: boolean;
	onChange: (newData: QuestionCategories[]) => void;
}

export function EnrollmentQuestions(props: EnrollmentQuestionsProps): React.ReactElement {
	const { studentQuestionCategories, editable, onChange } = props;

	const onChangeHandler = (changedQuestionCategoryIndex: number, changedQuestionIndex: number, newAnserValue: string): void => {
		const newQuestionCategories: QuestionCategories[] = _.cloneDeep(studentQuestionCategories);
		newQuestionCategories[changedQuestionCategoryIndex].questions[changedQuestionIndex].answer = newAnserValue;
		onChange(newQuestionCategories);
	};

	console.log(editable);

	return (
		<div>
			<List>
				{studentQuestionCategories.map((category, categoryIndex): React.ReactElement => {
					return (
						<div key={"category" + categoryIndex}>
							<Typography variant="h4" gutterBottom>
								{category.category}
							</Typography>

							{category.questions.map((question, questionIndex): React.ReactElement => {
								const [answer, setAnswer] = useState(question.answer);
								const debouncedAnswer = useDebounce<string>(answer, 500);

								useEffect(() => {
									onChangeHandler(categoryIndex, questionIndex, debouncedAnswer);
								}, [debouncedAnswer]);

								return (
									<div key={"question" + question.id + questionIndex}>
										<ListItem>
											<Grid container spacing={2} sx={{ justifyContent: "space-between" }}>
												<Grid item xs={4}>
													<div style={{ paddingRight: 20 }}>{question.question}</div>
												</Grid>

												<Grid item xs={6}>
													<TextField
														multiline
														maxRows={8}
														fullWidth
														value={answer}
														onChange={(e): void => {
															setAnswer(e.target.value);
														}}
													/>
												</Grid>
											</Grid>
										</ListItem>
										<Divider />
									</div>
								);
							})}
							{categoryIndex < studentQuestionCategories.length - 1 && <Divider />}
						</div>
					);
				})}
			</List>
		</div>
	);
}
