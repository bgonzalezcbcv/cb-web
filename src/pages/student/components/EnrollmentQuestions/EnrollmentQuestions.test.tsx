import React from "react";
import { act, getByText, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import * as UseDebounce from "../../../../hooks/useDebounce";
import EnrollmentQuestions from "./EnrollmentQuestions";
import { initialStudentData, expectedNewStudentData } from "./EnrollmentQuestions.fixture";

describe("EnrollmentQuestions", () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	test("should render correctly", async () => {
		const wrapper = render(
			<EnrollmentQuestions //
				student={initialStudentData}
				editable
				onChange={jest.fn}
				viewMode={"VIEW"}
			/>
		);

		const container = wrapper.container;

		expect(container).toMatchSnapshot();
	});

	test("should call onChange with answer", async () => {
		jest.spyOn(UseDebounce, "default").mockImplementation((answer: unknown): unknown => {
			return answer;
		});

		const onChangeSpy = jest.fn((): void => {
			return;
		});

		const wrapper = render(
			<EnrollmentQuestions //
				student={initialStudentData}
				editable
				onChange={onChangeSpy}
				viewMode={"VIEW"}
			/>
		);

		const answerTextField = getByText(wrapper.container, "answer 1a");

		act(() => {
			userEvent.type(answerTextField, "new");
		});

		expect(answerTextField.textContent).toBe("answer 1anew");

		expect(onChangeSpy).toHaveBeenCalledWith(expectedNewStudentData);
	});
});
