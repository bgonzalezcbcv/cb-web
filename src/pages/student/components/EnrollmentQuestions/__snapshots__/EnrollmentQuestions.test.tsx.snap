// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrollmentQuestions should render correctly 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div>
      <span
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
      >
        categoria 1
      </span>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1xu1mzw-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 1 pero es mucho muchomuchomuchomucho muchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomucho muchomuchomuchomuchomuchomuchomuchomuchomucho mas larga para romper todo
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 1a
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 2
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 2
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 3
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r2:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 3
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </div>
    <div>
      <span
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
      >
        categoria 2
      </span>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1xu1mzw-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 12
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r3:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 12
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 22
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r4:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 22
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-lg-6 MuiGrid-grid-xl-4 css-9h66bi-MuiGrid-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1ss44p5-MuiListItem-root"
          >
            <div
              class="MuiBox-root css-7n0sq1"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
                style="padding-right: 20px;"
              >
                question 32
              </span>
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r5:"
                    style="height: 0px; overflow: hidden;"
                  >
                    answer 32
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </li>
        </div>
      </div>
    </div>
  </ul>
</div>
`;
